src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs

```diff
@@ -0,0 +1,581 @@
+                                {
+                                    var alreadyAssociatedMethod = _generatedTypeToTypeArgumentInfo[generatedType].CreatingMethod;
+                                    logger?.LogWarning(new MessageOrigin(method), DiagnosticId.MethodsAreAssociatedWithUserMethod, method.GetDisplayName(), alreadyAssociatedMethod.GetDisplayName(), generatedType.GetDisplayName());
+                                }
+                                continue;
+                            }
+
+                            if (!CompilerGeneratedNames.IsLambdaOrLocalFunction(lambdaOrLocalFunction.Name))
+                                continue;
+
+                            if (isStateMachineMember)
+                            {
+                                if (method.OwningType is MetadataType owningType)
+                                {
+                                    callGraph.TrackCall(owningType, lambdaOrLocalFunction);
+                                }
+                            }
+                            else
+                            {
+                                callGraph.TrackCall(method, lambdaOrLocalFunction);
+                            }
+                        }
+                    }
+
+                    if (TryGetStateMachineType(method, out MetadataType? stateMachineType))
+                    {
+                        Debug.Assert(stateMachineType.ContainingType == type ||
+                            (CompilerGeneratedNames.IsGeneratedMemberName(stateMachineType.ContainingType.Name) &&
+                             stateMachineType.ContainingType.ContainingType == type));
+                        Debug.Assert(stateMachineType == stateMachineType.GetTypeDefinition());
+
+                        callGraph.TrackCall(method, stateMachineType);
+
+                        _compilerGeneratedTypeToUserCodeMethod ??= new Dictionary<MetadataType, MethodDesc>();
+                        if (!_compilerGeneratedTypeToUserCodeMethod.TryAdd(stateMachineType, method))
+                        {
+                            var alreadyAssociatedMethod = _compilerGeneratedTypeToUserCodeMethod[stateMachineType];
+                            logger?.LogWarning(new MessageOrigin(method), DiagnosticId.MethodsAreAssociatedWithStateMachine, method.GetDisplayName(), alreadyAssociatedMethod.GetDisplayName(), stateMachineType.GetDisplayName());
+                        }
+                        // Already warned above if multiple methods map to the same type
+                        // Fill in null for argument providers now, the real providers will be filled in later
+                        _generatedTypeToTypeArgumentInfo ??= new Dictionary<MetadataType, TypeArgumentInfo>();
+                        _ = _generatedTypeToTypeArgumentInfo.TryAdd(stateMachineType, new TypeArgumentInfo(method, null));
```

Comment:
Nit: it's odd to call TryAdd and then discard the result.

```suggestion
                        _generatedTypeToTypeArgumentInfo[stateMachineType] = new TypeArgumentInfo(method, null);
```